src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/SimpleChunkVisitor.java

```diff
@@ -0,0 +1,114 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.jenkinsci.plugins.workflow.graphanalysis;
+
+import org.jenkinsci.plugins.workflow.graph.FlowNode;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+
+/**
+ * This visitor's callbacks are invoked as we walk through a pipeline flow graph, and it splits it into chunks.
+ * <p></p> A {@link ForkScanner#visitSimpleChunks(SimpleChunkVisitor, ChunkFinder)} creates these FlowChunks using a {@link ChunkFinder} to define the chunk boundaries.
+ *
+ * <p></p> Implementations get to decide how to use & handle chunks.
+ * <p></p> <h3>At a minimum they should handle:</h3>
+ * <ul>
+ *     <li>Unbalanced numbers of chunk start/end calls (for incomplete flows)</li>
+ *     <li>A chunk end with no beginning (runs to start of flow, or never began)</li>
+ *     <li>A chunk start with no end (ex: a block that hasn't completed running)</li>
+ *     <li>Other starts/ends before we hit the closing one (nesting)</li>
+ *     <li>Atom nodes not within the current Chunk (visitor is responsible for handling state)</li>
+ * </ul>
+ *
+ * <em>Important implementation note: multiple callbacks can be invoked for a single node depending on its type.</em>
+ * <p></p>For example, we may capture parallels as chunks.
+ *
+ * <p></p><h3>Callbacks Reporting on chunk/parallel information:</h3>
+ * <ul>
+ *     <li>{@link #chunkStart(FlowNode, FlowNode, ForkScanner)} is called on the current node when we hit start of a boundary (inclusive) </li>
+ *     <li>{@link #chunkEnd(FlowNode, FlowNode, ForkScanner)} is called when we hit end of a boundary (inclusive)</li>
+ *     <li>{@link #atomNode(FlowNode, FlowNode, FlowNode, ForkScanner)} called when a node is neither start nor end.</li>
```

Comment:
@oleg-nenashev Hey now, leave my past work in nuclear physics out of it, I promise I don't glow green at night. ;) 

In all seriousness, AtomNode is a convention in pipeline itself (see: StepAtomNode).  This is following the standards there.